### ISSUE TITLE
MultiStreamProjection ignoring ShouldDelete

### ISSUE THREAD

**Issue by Wywarro** [JasperFx/marten#3162]
Hello!

I'm trying to upgrade Marten from 6.4.1 to 7.8 and some of the tests are failing.
Looks like it's ignoring ShouldDelete in MultiStreamProjection.
This example works super fine in 6.4.1

Here is my test:

```c#
public record TeamCreated(Guid TeamId);
public record TeamNameChanged(string TeamName);

public sealed class Team
{
    public Guid Id { get; set; }
    public string TeamName { get; set; } = "";

    internal Team()
    { }

    public void Apply(TeamCreated @event)
    {
        Id = @event.TeamId;
    }

    public void Apply(TeamNameChanged @event)
    {
        TeamName = @event.TeamName;
    }
}

public record MemberInvited(Guid TeamMembershipId, Guid TeamId, Guid UserIdInviting, Guid UserIdInvited);
public record MemberJoined;

public sealed class InvitationView
{
    public Guid Id { get; set; }
    public Guid TeamId { get; set; }
    public Guid UserIdInviting { get; set; }
    public string TeamName { get; set; } = "";
    public DateTimeOffset CreatedAt { get; set; }
}

public sealed class TeamInvitationGrouper : IAggregateGrouper<Guid>
{
    private readonly Type[] eventTypes =
    [
        typeof(MemberInvited),
        typeof(MemberJoined),
    ];

    public async Task Group(
        IQuerySession session,
        IEnumerable<IEvent> events,
        ITenantSliceGroup<Guid> grouping)
    {
        var eventsArray = events.ToArray();
        MembershipGroup(eventsArray, grouping);
        await TeamGroup(session, eventsArray, grouping);
    }

    private void MembershipGroup(
        IEnumerable<IEvent> events,
        ITenantSliceGroup<Guid> grouping)
    {
        var filteredEvents = events
            .Where(ev => eventTypes.Contains(ev.EventType))
            .GroupBy(e => e.StreamId)
            .Select(x => new
            {
                TeamMembershipId = x.Key,
                Events = x.ToArray(),
            })
            .ToList();

        foreach (var membershipEvents in filteredEvents)
        {
            grouping.AddEvents(membershipEvents.TeamMembershipId, membershipEvents.Events);
        }
    }

    private async Task TeamGroup(
        IQuerySession session,
        IEnumerable<IEvent> events,
        ITenantSliceGroup<Guid> grouping)
    {
        var teamEvents = events
            .OfType<IEvent<TeamNameChanged>>()
            .ToArray();

        if (teamEvents.Length == 0)
        {
            return;
        }

        var teamIds = teamEvents
            .Select(e => e.StreamId)
            .ToList();

        var result = await session.Query<InvitationView>()
            .Where(e => teamIds.Contains(e.TeamId))
            .Select(x => new { x.Id, x.TeamId })
            .ToListAsync();

        var eventsPerTeam = result.Select(g =>
            new
            {
                g.Id,
                Events = teamEvents.Where(ev => ev.StreamId == g.TeamId),
            });

        foreach (var group in eventsPerTeam)
        {
            grouping.AddEvents(group.Id, group.Events);
        }
    }
}

public sealed class InvitationProjection : MultiStreamProjection<InvitationView, Guid>
{
    public InvitationProjection()
    {
        IncludeType<TeamNameChanged>();
        IncludeType<MemberInvited>();
        IncludeType<MemberJoined>();

        CustomGrouping(new TeamInvitationGrouper());
    }

    public async Task<InvitationView> Create(
        IQuerySession session,
        IEvent<MemberInvited> @event)
    {
        var teamData = await session.Query<Team>()
            .Where(u => u.Id == @event.Data.TeamId)
            .Select(x => x.TeamName)
            .SingleOrDefaultAsync();

        return new InvitationView
        {
            Id = @event.Data.TeamMembershipId,
            TeamId = @event.Data.TeamId,
            UserIdInviting = @event.Data.UserIdInviting,
            TeamName = teamData ?? "",
            CreatedAt = @event.Timestamp,
        };
    }

    public InvitationView Apply(TeamNameChanged @event, InvitationView view)
    {
        view.TeamName = @event.TeamName;

        return view;
    }

    public bool ShouldDelete(MemberJoined @event)
    {
        return true;
    }
}

public sealed class MultiStreamProjectionIssueReproduction
{
    private readonly ITestOutputHelper testOutputHelper;

    private readonly Guid teamId = Guid.NewGuid();
    private readonly string teamName = "Yummy";

    private readonly Guid membershipId = Guid.NewGuid();
    private readonly Guid userId1 = Guid.NewGuid();
    private readonly Guid userId2 = Guid.NewGuid();

    public MultiStreamProjectionIssueReproduction(ITestOutputHelper testOutputHelper)
    {
        this.testOutputHelper = testOutputHelper;
    }

    private async Task<IDocumentStore> GiveMeStore(IMartenLogger logger)
    {
        var store = DocumentStore.For(options =>
        {
            options.Connection(ConnectionSource.ConnectionString);
            options.DatabaseSchemaName = ConnectionSource.SchemaName;
            options.Events.DatabaseSchemaName = ConnectionSource.SchemaName;

            options.AutoCreateSchemaObjects = AutoCreate.All;
            options.GeneratedCodeMode = TypeLoadMode.Dynamic;

            options.Projections.Snapshot<Team>(SnapshotLifecycle.Inline);
            options.Projections.Add<InvitationProjection>(ProjectionLifecycle.Async);

            options.Logger(logger);
        });

        await store.Advanced.Clean.DeleteAllDocumentsAsync();
        await store.Advanced.Clean.DeleteAllEventDataAsync();
        await store.Storage.ApplyAllConfiguredChangesToDatabaseAsync();

        return store;
    }

    [Fact]
    public async Task MethodName_ShouldDoWhat_WhenWhat()
    {
        var logger = new DogMeatLogger(testOutputHelper);
        var store = await GiveMeStore(logger);

        await using var session = store.LightweightSession();
        session.Logger = logger;

        session.Events.Append(teamId,
            new TeamCreated(teamId),
            new TeamNameChanged(teamName));

        await session.SaveChangesAsync();

        session.Events.Append(membershipId,
            new MemberInvited(membershipId,
                teamId,
                userId1,
                userId2));

        await session.SaveChangesAsync();

        using var daemon = await store.BuildProjectionDaemonAsync();
        await daemon.RebuildProjection<InvitationProjection>(CancellationToken.None);

        await AssertInvitation(session, membershipId, teamName);

        session.Events.Append(membershipId,
            new MemberJoined());

        await session.SaveChangesAsync();

        await daemon.RebuildProjection<InvitationProjection>(CancellationToken.None);
        await AssertInvitationDoesNotExist(session, membershipId);

        return;

        async Task AssertInvitation(IDocumentSession assertSession, Guid assertMembershipId, string assertTeamName)
        {
            var invitationReadModel = await assertSession.Query<InvitationView>()
                .SingleOrDefaultAsync(p => p.Id == assertMembershipId);

            invitationReadModel.Should().NotBeNull();
            invitationReadModel!.TeamName.Should().Be(assertTeamName);
        }

        async Task AssertInvitationDoesNotExist(IDocumentSession assertSession, Guid assertMembershipId)
        {
            var invitation = await assertSession.Query<InvitationView>()
                .SingleOrDefaultAsync(p => p.Id == assertMembershipId);

            invitation.Should().BeNull();
        }
    }
}
```

Test result:
```
Xunit.Sdk.XunitException
Expected invitation to be <null>, but found ProjectAlpha.InvitationView
{
    CreatedAt = <2024-04-25 12:47:05.598378 +2h>, 
    Id = {c3cd1a48-2757-463f-a054-04a7eb21fed6}, 
    TeamId = {7748d71f-4cb5-4b54-8764-b6591ad37d1e}, 
    TeamName = "Yummy", 
    UserIdInviting = {07e09053-df6c-4a96-8343-c46f97b626c0}
}.
```

**Comment by jeremydmiller** [comment]
@Wywarro Nothing to do with the projection itself or the multi-stream projection functionality (whew!). Issue for your test was in the daemon not checking for a new high water mark in between doing the first rebuild and the second rebuild. Easy money, this will be fixed in 7.9 in a bit.

**Comment by Wywarro** [comment]
Ow right!

7.9 is making all my tests green :) feels good ;)

Thanks for a quick action!
